File: ui/src/app/capture/timeline.ts
// Capture-timeline pure helpers (ADR-0013 §3.3, §4.4, §8). Owner: T-150. Scrub↔time mapping, the
// "reviewing N ago" wording and the shading arithmetic for a grid the backend measured — no DOM, no
// fetch, no RF constants, unit-tested.
//
// T-338 — **the band's span is not this file's to decide.** It was a hard-coded `WINDOW_S = 48 h`,
// which is the failure the user's invariant names: a scrubber sized from anything but the capture
// window offers times the IQ ring has already overwritten, and looks right while it lies. Every
// function here now takes `windowS` and there is no default, so nothing can silently fall back to a
// constant; `GET /api/timeline` supplies it from the ring's configured retention (docs/api.md,
// "The capture window, and the overview drawn on it").
//
// The band's *content* is not this file's either. Reducing a history grid to drawn columns used to
// happen here (`reduceActivity`, a max over every frequency cell plus a min/max normalisation);
// both were measurements, and T-334's rule puts measurements in the backend. `/api/timeline` now
// serves the compressed grid and its dynamic range, and what is left below is arithmetic over
// numbers the backend already decided.

/** The capture window `GET /api/timeline` reports: the IQ ring's configured retention (ADR-0014),
 * which is what the band spans — never the longer, lossy spectrum-history horizon. */
export interface CaptureWindow {
  /** Band start / end / span, Unix s. */
  t0S: number; t1S: number; spanS: number;
  /** What the ring currently holds, inside the band; `null` when it holds nothing. */
  buffered: { t0S: number; t1S: number } | null;
}

/** The `GET /api/timeline` response fields the band reads. */
export interface TimelineResponse {
  window?: {
    enabled?: boolean;
    retention_s?: number | null;
    t0_s?: number | null; t1_s?: number | null; span_s?: number | null;
    buffered?: { t0_s: number; t1_s: number } | null;
  } | null;
  grid?: OverviewResponse | null;
}

/**
 * The capture window, or `null` when the server reports none — no ring, no retention, or no live
 * edge yet. `null` is **unknown**, and the band must draw it as unknown: a default span here would
 * be the 48 h constant all over again.
 */
export function captureWindow(r: TimelineResponse | null): CaptureWindow | null {
  const w = r?.window;
  if (!w) return null;
  const { t0_s, t1_s, span_s } = w;
  if (typeof t0_s !== "number" || typeof t1_s !== "number" || typeof span_s !== "number") return null;
  if (!(span_s > 0) || !(t1_s > t0_s)) return null;
  const b = w.buffered;
  return {
    t0S: t0_s, t1S: t1_s, spanS: span_s,
    buffered: b && typeof b.t0_s === "number" && typeof b.t1_s === "number" ? { t0S: b.t0_s, t1S: b.t1_s } : null,
  };
}

/** The compressed overview grid `GET /api/timeline` draws the band from: `nt × nf` cells, row-major
 * (time then frequency), already folded onto the capture window by the backend. `null` in `max_db`
 * is **not observed**, never quiet (C26). */
export interface OverviewResponse {
  nt: number; nf: number;
  max_db: readonly (number | null)[];
  range_db?: { lo: number; hi: number } | null;
  /** Cells in the grid, and the ones something was folded into. */
  cells?: number; observed_cells?: number;
}

/** The observed fraction of the capture window, from the grid's own counts; `null` when no grid has
 * been served (unknown, never "nothing was observed"). */
export function observedFraction(grid: OverviewResponse | null | undefined): number | null {
  if (!grid || typeof grid.cells !== "number" || typeof grid.observed_cells !== "number") return null;
  return grid.cells > 0 ? grid.observed_cells / grid.cells : null;
}

/**
 * The 0–1 shade of one overview cell, against the range **the backend measured** (`range_db`), or
 * `null` for a cell nothing was observed in.
 *
 * Deliberately not a normalisation over the values in hand: deciding a band's dynamic range from
 * whatever came back is a measurement, and it belongs where the floor and the occupancy are known.
 * With no range served there is nothing measured to shade against, so every cell reads unknown.
 */
export function overviewShade(grid: OverviewResponse, index: number): number | null {
  const v = grid.max_db[index];
  const r = grid.range_db;
  if (v === null || v === undefined || !r) return null;
  const span = r.hi > r.lo ? r.hi - r.lo : 1;
  return Math.max(0, Math.min(1, (v - r.lo) / span));
}

export interface Scrub { live: boolean; tS: number; agoS: number }

/** Maps a pointer's percent along the band to a time cursor: > 98.5 % is live, otherwise it's
 * `agoS` seconds before `nowS` (0…`windowS`, clamped). `windowS` is the capture window's span and
 * has no default — see the note at the top of this file. */
export function scrubToTime(pct: number, nowS: number, windowS: number): Scrub {
  const clamped = Math.max(0, Math.min(100, pct));
  const live = clamped > 98.5;
  const agoS = live ? 0 : ((100 - clamped) / 100) * windowS;
  return { live, tS: nowS - agoS, agoS };
}

/** The percent along the band for a given "ago" duration (the inverse of `scrubToTime`, for
 * drawing the playhead while reviewing). */
export function pctForAgo(agoS: number, windowS: number): number {
  return Math.max(0, Math.min(100, 100 - (agoS / windowS) * 100));
}

/** "reviewing N ago" wording: minutes under an hour, else hours to one decimal. */
export function agoText(agoS: number): string {
  const mins = Math.round(agoS / 60);
  return mins < 60 ? `${Math.max(0, mins)} min` : `${(mins / 60).toFixed(1)} h`;
}

/** How much of the capture window was observed, from the overview grid the backend served: its own
 * `observed_cells / cells`, over the window it says it spans. `null` (not asked yet, or no window)
 * reads as unknown, never as "nothing observed". */
export function coverageText(observedFraction: number | null, windowS: number | null): string {
  if (observedFraction === null || windowS === null) return "coverage unknown";
  return `${Math.round(observedFraction * 100)}% of the retained ${durationText(windowS)} observed`;
}

/** A capture window's length in words. Unlike `agoText` this keeps seconds, because a retention is
 * commonly seconds (`--iq-retention 90s`) and rounding it to "2 min" would misstate the span the
 * band is claiming to be. */
export function durationText(s: number): string {
  if (s < 90) return `${Math.round(s)} s`;
  if (s < 3600) return `${Math.round(s / 60)} min`;
  return `${(s / 3600).toFixed(1)} h`;
}

/** The current view span to query `/api/history` for: the live geometry once T-152 sets it,
 * else a fallback from the device's tuned centre/rate (both from the store, no analysis). */
export function currentSpan(input: {
  live: { loHz: number; hiHz: number } | null;
  device: { centerHz: number | null; sampleRateHz: number | null };
}): { loHz: number; hiHz: number } | null {
  if (input.live) return input.live;
  const { centerHz, sampleRateHz } = input.device;
  if (centerHz === null || sampleRateHz === null) return null;
  return { loHz: centerHz - sampleRateHz / 2, hiHz: centerHz + sampleRateHz / 2 };
}

/**
 * A key that changes exactly when [[currentSpan]] does (T-386): the band the capture panel's
 * **content** — its overview grid, its coverage fraction and gaps, its event marks — is an answer
 * about.
 *
 * The panel polls once a minute. Before T-386 it read the span *inside* that poll, so a retune or a
 * zoom left up to 60 s of the **previous band's** overview, coverage and marks on screen, labelled
 * as this window's. That is worse than an empty surface: it is the window rule broken by rendering
 * data that exists for some *other* window as if it were this one. Tagging every answer with the
 * band it came from lets the panel discard it the instant the view leaves that band, rather than
 * drawing it until the next tick.
 *
 * Rounded to whole hertz so a float that re-derives to the same tuning does not look like a move.
 */
export function bandKey(span: { loHz: number; hiHz: number } | null): string {
  return span ? `${Math.round(span.loHz)}/${Math.round(span.hiHz)}` : "";
}

// ---- time-window select (T-194) ----

/** Pointer travel (px) that turns a scrub into a deliberate time-window drag (mirrors the live
 * view's DRAG_PX rule, ADR-0013 §5, centre/overlays.ts). */
export const DRAG_PX = 6;

export interface TimeWindow { t_lo: number; t_hi: number }

/**
 * The `[t_lo, t_hi]` a drag from percent `pctA` to `pctB` along the band selects, chronologically
 * ordered (through `scrubToTime`'s mapping, so the LIVE edge resolves to `nowS`); null when it has
 * no width (both ends resolved to the same instant — e.g. both past the LIVE threshold).
 */
export function timeWindowFromScrub(pctA: number, pctB: number, nowS: number, windowS: number): TimeWindow | null {
  const a = scrubToTime(pctA, nowS, windowS), b = scrubToTime(pctB, nowS, windowS);
  const t_lo = Math.min(a.tS, b.tS), t_hi = Math.max(a.tS, b.tS);
  return t_hi > t_lo ? { t_lo, t_hi } : null;
}

/** A default name for a time-window selection: its UTC wall-clock span, to the second. */
export function timeRegionName(t_lo: number, t_hi: number): string {
  const f = (t: number) => `${new Date(t * 1000).toISOString().slice(11, 19)}Z`;
  return `${f(t_lo)}–${f(t_hi)}`;
}

// ---- past events on the scrubber, and what backs a scrub-back (T-263, ADR-0017 TM-7) ----
//
// The marks are the durable events a user scrubs *to*. They are placed from timespans the API
// already serves per row — `presence.last_interval` (the latest interval intersecting the request's
// window) and `recurrence.recent[]` (its recent appearances) — and nothing here decides which rows
// qualify or what state they are in: that is the backend's predicate (thin-client rule, §1).
//
// Deliberately no liveness on a mark. `presence.last_interval` carries `open`; a `recurrence`
// appearance carries no such field, so a mark that claimed one for both would be asserting
// something about half of them that was never measured. A mark says only *an event happened over
// this timespan*; liveness is the list's answer, from `presence.liveness`.

/** The `/api/inventory` row fields the scrubber's marks are placed from. */
export interface EventRow {
  id: string;
  state: string;
  presence?: { last_interval: { t_start_s: number; t_end_s: number } | null };
  recurrence?: { recent?: readonly { t_start_s: number; t_end_s: number }[] } | null;
}

export interface EventMark {
  id: string; state: "candidate" | "confirmed";
  leftPct: number; widthPct: number;
  tStartS: number; tEndS: number;
}

/** Narrowest a mark is drawn, % of the band. A one-off burst is milliseconds against a band of
 * minutes, so its true width rounds to nothing; it is still the event this whole model exists to make
 * first-class (ADR-0017 §1.2), so it is drawn at this floor. The floor is a **drawing** minimum —
 * `tStartS`/`tEndS` keep the measured timespan, and nothing reads the width back as a duration. */
export const MIN_MARK_PCT = 0.3;

/** Most marks drawn at once. A render cap, never a claim about what happened: the newest are kept,
 * and History (TM-8) is where the complete catalogue lives. */
export const MAX_MARKS = 400;

/**
 * Past events placed on the retained band, newest last. One mark per distinct timespan a row
 * reports — its `presence.last_interval` and each `recurrence.recent[]` appearance, de-duplicated
 * where they name the same span. A timespan that ended before the retained window, or that has not
 * started, is omitted rather than clamped to the band's edge (which would put an event at a time it
 * never happened). Rows in neither list are skipped.
 */
export function eventMarks(rows: readonly EventRow[], nowS: number, windowS: number): EventMark[] {
  const out: EventMark[] = [];
  for (const r of rows) {
    if (r.state !== "candidate" && r.state !== "confirmed") continue;
    const state = r.state;
    const spans = [...(r.recurrence?.recent ?? [])];
    const last = r.presence?.last_interval;
    if (last) spans.push(last);
    const seen = new Set<string>();
    for (const s of spans) {
      if (![s.t_start_s, s.t_end_s].every(Number.isFinite)) continue;
      const key = `${s.t_start_s}|${s.t_end_s}`;
      if (seen.has(key)) continue;
      seen.add(key);
      if (nowS - s.t_end_s > windowS || s.t_start_s > nowS) continue;
      const leftPct = pctForAgo(nowS - s.t_start_s, windowS), rightPct = pctForAgo(nowS - s.t_end_s, windowS);
      out.push({
        id: r.id, state,
        leftPct: Math.min(leftPct, 100 - MIN_MARK_PCT),
        widthPct: Math.max(MIN_MARK_PCT, rightPct - leftPct),
        tStartS: s.t_start_s, tEndS: s.t_end_s,
      });
    }
  }
  out.sort((a, b) => a.tEndS - b.tEndS);
  return out.length > MAX_MARKS ? out.slice(out.length - MAX_MARKS) : out;
}

/** A mark's hover text: when it ended and how long it lasted — the measured timespan, never the
 * drawn width. A sub-second event reads in milliseconds rather than rounding to "0 s". */
export function eventMarkTitle(m: Pick<EventMark, "tStartS" | "tEndS">, nowS: number): string {
  const dur = Math.max(0, m.tEndS - m.tStartS);
  const lasted = dur < 1 ? `${Math.round(dur * 1000)} ms` : dur < 60 ? `${dur.toFixed(1)} s` : agoText(dur);
  return `${agoText(Math.max(0, nowS - m.tEndS))} ago · lasted ${lasted}`;
}

// ---- what backs a scrub-back: the IQ ring, and whether anything was observed at all ----
//
// Two different "nothing here" answers that must never be collapsed. *Nothing was on the air* is a
// measurement. *No data for this window* is the absence of one — the receiver was not listening, or
// the IQ ring has already rolled past it. The user acts differently on each, so neither is ever
// rendered as the other, and an unanswered poll reads as **unknown**, not as either.

/** One `coverage_summary.gaps[]` run of `GET /api/history`: a stretch in which **no** cell of the
 * grid was observed. "A gap is never reported as quiet." */
export interface CoverageGap { t0_s: number; t1_s: number }

/**
 * What the ring **holds** placed on the band (ADR-0014), or `null` when it holds nothing or no
 * window has been answered — both "unknown", never "the ring is empty".
 *
 * T-338: this sits *inside* the band rather than sizing it. The band is the ring's configured
 * retention; a ring part-way through filling covers only part of it, and that difference is the
 * thing this track exists to show.
 */
export function bufferedSpan(w: CaptureWindow | null): SelSpan | null {
  if (!w?.buffered) return null;
  const { t0S, t1S } = w.buffered;
  if (!(t1S > t0S)) return null;
  const leftPct = pctForAgo(w.t1S - t0S, w.spanS), rightPct = pctForAgo(w.t1S - t1S, w.spanS);
  return rightPct > leftPct ? { id: "ring", leftPct, widthPct: rightPct - leftPct } : null;
}

/** Whether the reviewed instant still has IQ behind it: `"live"` at the live edge, `"ring"` inside
 * what the ring holds, `"outside-ring"` past it, `"unknown"` when no window has been answered. */
export type IqBacking = "live" | "ring" | "outside-ring" | "unknown";

export function iqBackingAt(tS: number, live: boolean, w: CaptureWindow | null): IqBacking {
  if (live) return "live";
  if (!w?.buffered) return "unknown";
  return tS >= w.buffered.t0S && tS <= w.buffered.t1S ? "ring" : "outside-ring";
}

/** Whether the reviewed instant falls in a stretch the history grid reports as observed by nothing;
 * `null` when no coverage summary has been answered (unknown, not "observed"). */
export function inCoverageGap(tS: number, gaps: readonly CoverageGap[] | null): boolean | null {
  return gaps === null ? null : gaps.some((g) => tS >= g.t0_s && tS <= g.t1_s);
}

/**
 * What the scrubbed instant is backed by, in one clause for the capture note. Empty while live (the
 * band's own coverage text already speaks for the live edge).
 *
 * An unobserved stretch wins over everything else: it is the one case where an empty Candidate list
 * means *the receiver was not listening*, and saying anything else there would let silence read as
 * a measurement.
 */
export function scrubDataNote(tS: number, live: boolean, w: CaptureWindow | null, gaps: readonly CoverageGap[] | null): string {
  if (live) return "";
  if (inCoverageGap(tS, gaps)) return "nothing was observed here — no data for this window, not a quiet band";
  switch (iqBackingAt(tS, live, w)) {
    case "ring": return "IQ retained for this window";
    case "outside-ring": return "past the IQ ring — lists come from stored history, no IQ detail";
    default: return "IQ coverage unknown";
  }
}

export interface SelSpan { id: string; leftPct: number; widthPct: number }

/**
 * Time-windowed selections placed on the retained band (the inverse of `scrubToTime`, via
 * `pctForAgo`). A selection with no time window, or whose window falls wholly outside the
 * retained window, is omitted.
 */
export function selectionSpans(list: readonly { id: string; t_lo?: number; t_hi?: number }[], nowS: number, windowS: number): SelSpan[] {
  const out: SelSpan[] = [];
  for (const s of list) {
    if (s.t_lo === undefined || s.t_hi === undefined) continue;
    const leftPct = pctForAgo(nowS - s.t_lo, windowS), rightPct = pctForAgo(nowS - s.t_hi, windowS);
    if (rightPct > leftPct) out.push({ id: s.id, leftPct, widthPct: rightPct - leftPct });
  }
  return out;
}

File: crates/hk-api/src/timeline.rs
//! `GET /api/timeline` — the capture window, and the compressed overview drawn on it (T-338).
//!
//! From the user's invariant (CLAUDE.md, "Time, the waterfall, and the live view"):
//!
//! > **The timeline is the capture window, and it is a visualization.** The scrubbable
//! > capture-history timeline spans **exactly the configured recording/retention duration** — no
//! > more (not the longer, lossy spectrum-history retention), no less — and grows or shrinks when
//! > that duration is reconfigured. It is itself a data display: a **compressed "sideways"
//! > overview waterfall** of the retained capture (time along its long axis), never an empty box.
//!
//! Two things had to become the backend's, and this route is both of them.
//!
//! # 1. The horizon is the ring's, not the history's
//!
//! There are two retention horizons on this server and they are **deliberately different lengths**:
//! the IQ capture ring ([ADR-0014], `--iq-retention`, minutes) and the tiered spectrum-history
//! pyramid (lossy, byte-budgeted, days). Sizing the scrubber from the pyramid — or, as the client
//! did, from a hard-coded 48 h constant — makes it **promise capture that no longer exists**: the
//! user can scrub to a time the ring overwrote long ago, and the band looks right while it lies.
//! So `window` here is the ring's, `window.horizon` says so in the response, and
//! `GET /api/navigation`'s `time.latest_s` (the history horizon) is not consulted for it.
//!
//! The span served is `retention_s`, the **configured** window, not `t1 − t0` of what the ring
//! currently holds: a ring five minutes into a one-hour retention is a mostly-empty capture window,
//! not a five-minute one, and the timeline that shrank to fit it would grow under the user as the
//! ring filled. What the ring holds is reported beside it as `window.buffered`, so the client can
//! draw the difference (T-263's ring track) instead of inferring it.
//!
//! # 2. The overview is a measurement, so it is made here
//!
//! The band is a data display, which means something has to decide what each drawn cell shows.
//! T-334 put that decision in the backend: mapping a time to a pixel is presentation, *choosing
//! which value represents an interval* is a measurement. The pyramid alone cannot serve this
//! particular picture — its ladder couples the axes, so the tier with cells coarse enough in
//! frequency for a thin strip's rows (100 kHz) has one-day time cells — so the tier is read at the
//! timeline's own time resolution and folded onto `columns × rows` by
//! [`hk_store::RegionHistory::overview`], which carries only statistics that fold **exactly**
//! (max-hold, peak occupancy, summed frames, mean coverage). The grid's own dynamic range is served
//! too: picking a colour scale from the numbers you happen to hold is a measurement as well.
//!
//! `resolution` is T-334's block, with T-341's three-valued [`DetailSource`]: the timeline draws
//! the ring's window but its pixels come from the **pyramid**, so it claims `spectrum-history`
//! (or `survey-overview` for a span no capture window could hold) and never `live-iq`.
//!
//! | Method | Path | Query | Answers |
//! |---|---|---|---|
//! | GET | `/api/timeline` | `?[f_lo=&f_hi=][&columns=1..4096, default 96][&rows=1..512, default 1]` | `{window, region, grid, resolution}` |
//!
//! [ADR-0014]: ../../../docs/adr/0014-iq-capture-ring.md

use serde_json::{Value, json};

use crate::http::ApiState;
use crate::navigation::DetailSource;
use crate::query::{ApiError, Params, Region, count, parse_freq_only};

/// Columns the overview is drawn in when the caller names none.
pub const DEFAULT_COLUMNS: usize = 96;
/// Most columns accepted (a band cannot draw more than this, and each is a measurement).
pub const MAX_COLUMNS: usize = 4096;
/// Frequency rows when the caller names none: a single-row activity strip.
pub const DEFAULT_ROWS: usize = 1;
/// Most frequency rows accepted.
pub const MAX_ROWS: usize = 512;

/// The capture window the timeline spans, read from the IQ ring's own status.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct CaptureWindow {
    /// Whether a ring is running.
    pub enabled: bool,
    /// The configured retention, s — **the band's span, exactly**. `None` when no ring exists here
    /// to state one.
    pub retention_s: Option<f64>,
    /// The live edge of capture, Unix s.
    pub t1_s: Option<f64>,
    /// What the ring currently holds, Unix s; `None` when it holds nothing.
    pub buffered: Option<(f64, f64)>,
}

impl CaptureWindow {
    /// The band: `[t1 − retention, t1]`. `None` unless both a retention and a live edge are known —
    /// a timeline with no capture window behind it is drawn as unknown, never as a default span.
    pub fn band(&self) -> Option<(f64, f64)> {
        match (self.retention_s, self.t1_s) {
            (Some(r), Some(t1)) if r > 0.0 && t1.is_finite() => Some((t1 - r, t1)),
            _ => None,
        }
    }
}

fn f64_of(v: &Value, key: &str) -> Option<f64> {
    v.get(key).and_then(Value::as_f64).filter(|x| x.is_finite())
}

/// Reads the ring's status (`limit=1`: the window is wanted, not the segment list) and the history's
/// live edge, which stands in for `t1` only while the ring holds nothing yet.
pub(crate) fn capture_window(state: &ApiState) -> (CaptureWindow, Value) {
    let status = state.iq_buffer.as_deref().map(|c| {
        c.status(&crate::iqbuffer::IqBufferQuery {
            t0: None,
            t1: None,
            limit: 1,
        })
    });
    let Some(s) = status else {
        return (
            CaptureWindow {
                enabled: false,
                retention_s: None,
                t1_s: None,
                buffered: None,
            },
            // No ring at all on this server: the reason there is no capture window.
            json!("no IQ capture buffer on this server"),
        );
    };
    let buffered = match (f64_of(&s, "t0"), f64_of(&s, "t1")) {
        (Some(a), Some(b)) if b > a => Some((a, b)),
        _ => None,
    };
    // The live edge: the ring's newest sample, else the history's newest frame. Never wall-clock —
    // a replay or a time-compressed scene runs on its own clock (T-125), and a band anchored to
    // `now` would place its capture in the future.
    let t1_s = buffered.map(|(_, b)| b).or_else(|| {
        crate::http::with_history(state, |p| {
            Ok(p.latest_frame_end().map(|t| t.as_unix_nanos() as f64 / 1e9))
        })
        .ok()
        .flatten()
    });
    (
        CaptureWindow {
            enabled: s.get("enabled").and_then(Value::as_bool).unwrap_or(false),
            retention_s: f64_of(&s, "retention_s"),
            t1_s,
            buffered,
        },
        s.get("reason").cloned().unwrap_or(Value::Null),
    )
}

fn window_json(w: &CaptureWindow, reason: &Value) -> Value {
    let band = w.band();
    json!({
        // Which retention this span is. The spectrum-history horizon is longer and lossy; a
        // scrubber sized from it offers times the ring has already overwritten.
        "horizon": "iq-ring",
        "enabled": w.enabled,
        "reason": reason.clone(),
        "retention_s": w.retention_s,
        "t0_s": band.map(|(a, _)| a),
        "t1_s": band.map(|(_, b)| b),
        // Equal to `retention_s` by construction, and served so a client never computes it.
        "span_s": band.map(|(a, b)| b - a),
        "buffered": w.buffered.map(|(a, b)| json!({"t0_s": a, "t1_s": b, "span_s": b - a})),
    })
}

fn grid_json(o: &hk_store::Overview) -> Value {
    let num = |x: f32| -> Value { if x.is_finite() { json!(x) } else { Value::Null } };
    json!({
        "nt": o.nt,
        "nf": o.nf,
        "t0_s": o.t0_ns as f64 / 1e9,
        "t_cell_s": o.t_cell_ns / 1e9,
        "f_lo_hz": o.f_lo_hz,
        "f_cell_hz": o.f_cell_hz,
        // Row-major, time then frequency. `null` is **not observed**, never quiet (C26).
        "max_db": Value::Array(o.cells.iter().map(|c| num(c.max_db)).collect()),
        "occupancy_max": Value::Array(o.cells.iter().map(|c| num(c.occupancy_max)).collect()),
        "coverage": Value::Array(o.cells.iter().map(|c| json!(c.coverage)).collect()),
        "frames": Value::Array(o.cells.iter().map(|c| json!(c.frames)).collect()),
        "cells": o.cells.len(),
        "observed_cells": o.observed_cells,
        // The grid's own observed range of `max_db`, so the client draws a scale it was given
        // rather than deciding one from the values it happens to hold. `null` = nothing observed.
        "range_db": o.range_db.map(|(lo, hi)| json!({"lo": lo, "hi": hi})),
    })
}

fn resolution_json(
    r: &crate::query::OverviewRead,
    region: &Region,
    columns: usize,
    rows: usize,
    levels: usize,
    max_live_span_hz: Option<f64>,
) -> Value {
    let served_span_hz = region.freq.width_hz();
    // The timeline's *horizon* is the ring's; its *pixels* are the pyramid's. So it never claims
    // `live-iq`, exactly as `/api/history` does not.
    let source = match crate::navigation::live_window_verdict(served_span_hz, max_live_span_hz) {
        DetailSource::LiveIq => DetailSource::SpectrumHistory,
        other => other,
    };
    json!({
        "source": source.as_str(),
        "live": source.is_live(),
        "statement": source.statement(),
        // Which retention the picture spans, beside which tier drew it: the two questions the
        // scrubber has to answer honestly, and they have different answers.
        "horizon": "iq-ring",
        "served_span_hz": served_span_hz,
        "max_live_span_hz": max_live_span_hz,
        "level": r.level,
        "levels": levels,
        // The drawn cell, and the tier cell it was folded from. When `src_t_cell_s` is the larger
        // of the two, one measured value repeats across columns (T-334's safe direction, stated).
        "t_cell_s": r.grid.t_cell_ns / 1e9,
        "f_cell_hz": r.grid.f_cell_hz,
        "src_t_cell_s": r.src_t_cell_s,
        "src_f_cell_hz": r.src_f_cell_hz,
        "requested": { "columns": columns, "rows": rows },
        "served": { "nt": r.grid.nt, "nf": r.grid.nf, "cells": r.grid.nt * r.grid.nf },
        // The fold happens here, so the grid is always exactly the shape asked for: unlike
        // `/api/history`, this route never hands back more cells than the view can draw.
        "reduced_from": { "nt": r.grid.src_nt, "nf": r.grid.src_nf },
        "matched": true,
        "over_resolved": Value::Array(Vec::new()),
    })
}

/// `GET /api/timeline`.
pub fn timeline_json(state: &ApiState, q: &Params) -> Result<Value, ApiError> {
    const ALLOWED: [&str; 5] = ["f_lo", "f_hi", "columns", "rows", "token"];
    if let Some((k, _)) = q.iter().find(|(k, _)| !ALLOWED.contains(&k.as_str())) {
        return Err(ApiError::new(
            400,
            format!("unknown parameter {k:?} (allowed: f_lo, f_hi, columns, rows)"),
        ));
    }
    let columns = count(q, "columns", DEFAULT_COLUMNS, MAX_COLUMNS)?;
    let rows = count(q, "rows", DEFAULT_ROWS, MAX_ROWS)?;
    let freq = parse_freq_only(q)?;
    let (w, reason) = capture_window(state);
    let mut out = json!({
        "window": window_json(&w, &reason),
        "region": freq.map(|f| json!({"lo_hz": f.lo_hz, "hi_hz": f.hi_hz})),
        "grid": Value::Null,
        "resolution": Value::Null,
    });
    // No band or no region means no picture — and that is said as `null`, not drawn as an empty
    // box with a plausible span. There is nothing to be honest about when there is no window.
    let (Some((t0_s, t1_s)), Some(f)) = (w.band(), freq) else {
        return Ok(out);
    };
    let region = Region {
        freq: f,
        t0_ns: (t0_s * 1e9).round() as i64,
        t1_ns: (t1_s * 1e9).round() as i64,
    };
    let max_live = crate::http::max_live_span_hz(state);
    let (grid, resolution) = crate::http::with_history(state, |p| {
        let r = crate::query::overview_read(p, &region, columns, rows)?;
        let levels = p.geometry().n_levels();
        Ok((
            grid_json(&r.grid),
            resolution_json(&r, &region, columns, rows, levels, max_live),
        ))
    })?;
    if let Some(obj) = out.as_object_mut() {
        obj.insert("grid".into(), grid);
        obj.insert("resolution".into(), resolution);
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn w(retention_s: Option<f64>, t1_s: Option<f64>) -> CaptureWindow {
        CaptureWindow {
            enabled: true,
            retention_s,
            t1_s,
            buffered: None,
        }
    }

    #[test]
    fn the_band_is_the_configured_retention_not_what_the_ring_holds() {
        // A ring five minutes into a one-hour retention is a mostly-empty one-hour capture window.
        // Sizing the band to what it holds would make the timeline grow under the user.
        let mut c = w(Some(3600.0), Some(1_789_300_920.0));
        c.buffered = Some((1_789_300_620.0, 1_789_300_920.0));
        let (t0, t1) = c.band().unwrap();
        assert_eq!(t1 - t0, 3600.0);
        assert_eq!(t1, 1_789_300_920.0);
    }

    #[test]
    fn the_band_grows_and_shrinks_with_the_retention() {
        let edge = 1_789_300_920.0;
        let span = |r: f64| {
            let (a, b) = w(Some(r), Some(edge)).band().unwrap();
            b - a
        };
        assert_eq!(span(120.0), 120.0);
        assert_eq!(span(3600.0), 3600.0);
        assert_eq!(span(30.0), 30.0);
    }

    #[test]
    fn no_retention_or_no_live_edge_is_no_band_rather_than_a_default_one() {
        assert_eq!(w(None, Some(1.0)).band(), None);
        assert_eq!(w(Some(120.0), None).band(), None);
        assert_eq!(w(Some(0.0), Some(1.0)).band(), None);
    }
}
